### PR TITLE
Add barrierDismissable argument to showDialog

### LINKED ISSUE #8354: Add barrierDismissable to showDialog
I am trying to test the WillPopScope here is the example code:
```dart
showDialog(context: context,
              child: new WillPopScope(
                  onWillPop: () async => false,
                  child: new AlertDialog(
                      title: new Text("Pass!"),
                      content: new Markdown(
                          data: "You Pass"),
                      actions: [
                        new FlatButton(
                            child: new Text("Continue"),
                            onPressed: () {
                              goto(nextPage);
                            })
                      ]
                  ))
          );
```
The expected behavior is that I can't dismiss the dialog. Yet I can
From debugging here is the route being set in WillPopScopeState

```
_DialogRoute("null", animation: AnimationController(▶ 0.000; for _DialogRoute))
```

### PR DESCRIPTION
**Pull request by abarth**
In some situations, dialogs don't want the barrier to be dismissable.

Fixes #8354

### CODE REVIEW

**Comment by abarth** [comment]
@HansMuller @xster

File: packages/flutter/lib/src/material/dialog.dart [hunk 1]
```diff
@@ -316,9 +316,12 @@ class SimpleDialog extends StatelessWidget {
 
 class _DialogRoute<T> extends PopupRoute<T> {
   _DialogRoute({
-    this.child,
-    this.theme,
-  });
+    @required this.child,
+    @required this.theme,
+    @required bool barrierDismissable,
```
Comment (xster) [inline packages/flutter/lib/src/material/dialog.dart]: What about just defaulting to false?

**Review by xster — APPROVED** [review]

File: packages/flutter/lib/src/material/dialog.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter/lib/src/material/dialog.dart]: Good idea.

File: packages/flutter/lib/src/material/dialog.dart [hunk 2]
```diff
@@ -316,9 +316,12 @@ class SimpleDialog extends StatelessWidget {
 
 class _DialogRoute<T> extends PopupRoute<T> {
   _DialogRoute({
-    this.child,
-    this.theme,
-  });
+    @required this.theme,
+    bool barrierDismissable: true,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dialog.dart]: doesn't matter what the default is, this is only called from one place and that place lists the argument.

Comment (abarth) [inline packages/flutter/lib/src/material/dialog.dart]: That's true, but we might as well have ergonomic private APIs.

**Pull request merged**